### PR TITLE
IGNITE-17426 C++ 3.0: Implement table API

### PR DESCRIPTION
**Pull request by isapego**
- Implemented `Tables::getTablesAsync` method;
- Re-factored `IgniteResult` class;
- Re-factored `protocol::Reader` class;
- Re-factored tests.

### CODE REVIEW

File: modules/platforms/cpp/client/include/ignite/table/table.h [hunk 1]
```diff
@@ -38,8 +38,8 @@ class Table {
 
 public:
     // Deleted
-    Table(const Table &) = delete;
-    Table &operator=(const Table &) = delete;
+    Table(const Table &) = default;
+    Table &operator=(const Table &) = default;
```
Comment (ademakov) [inline modules/platforms/cpp/client/include/ignite/table/table.h:42]: Now the comment above is false.

File: modules/platforms/cpp/client/src/cluster_connection.h [hunk 2]
```diff
@@ -97,7 +97,7 @@ class ClusterConnection : public std::enable_shared_from_this<ClusterConnection>
      * @tparam T Result type.
      * @param op Operation code.
      * @param wr Writer function.
-     * @param rd Reader function.
+     * @param handler Response handler.
      * @return Future result.
```
Comment (ademakov) [inline modules/platforms/cpp/client/src/cluster_connection.h:101]: There is no return value here.

File: modules/platforms/cpp/client/src/cluster_connection.h [hunk 3]
```diff
@@ -114,6 +114,19 @@ class ClusterConnection : public std::enable_shared_from_this<ClusterConnection>
         }
     }
 
+    /**
+     * Perform request.
+     *
+     * @tparam T Result type.
+     * @param op Operation code.
+     * @param handler Response handler.
+     * @return Future result.
```
Comment (ademakov) [inline modules/platforms/cpp/client/src/cluster_connection.h:123]: There is no return value here.

File: modules/platforms/cpp/protocol/include/ignite/protocol/reader.h [hunk 4]
```diff
@@ -87,11 +89,46 @@ class Reader {
...
+     * @tparam V Value type.
+     * @param handler Pair handler.
+     */
+     template<typename K, typename V>
+     void readMap(const std::function<void(K&&, V&&)>& handler) {
+        checkDataInStream();
+
+        if (m_currentVal.data.type != MSGPACK_OBJECT_MAP)
+            throw ignite_error("The value in stream is not a Map");
+
```
Comment (ademakov) [inline modules/platforms/cpp/protocol/include/ignite/protocol/reader.h:125]: Looks like  `readMapSize()` can be used here.

**Review by ademakov — COMMENTED** [review]
LGTM

File: modules/platforms/cpp/client/src/table/tables_impl.h [hunk 5]
```diff
@@ -47,10 +47,19 @@ class TablesImpl {
...
      * @return TableImpl with corresponding name.
-     * @throw ignite_error In case of error.
+     * @throw ignite_error In case of error while trying to send a request.
      */
     void getTableAsync(const std::string &name, ignite_callback<std::optional<Table>> callback);
 
+    /**
+     * Gets all tables.
+     *
+     * @param callback Callback to be called once operation is complete. On success callback called with a vector of
```
Comment (ptupitsyn) [inline modules/platforms/cpp/client/src/table/tables_impl.h:57]:
```suggestion
     * @param callback Callback to be called once operation is complete. On success, the callback is invoked with a vector of
```

**Review by ptupitsyn — APPROVED** [review]

File: modules/platforms/cpp/common/factory.h [hunk 6]
```diff
@@ -42,4 +42,26 @@ class factory {
     virtual std::unique_ptr<T> build() = 0;
 };
 
+/**
+ * Basic factory class.
+ *
+ * @tparam TB Base type.
+ * @tparam TC Concreate type.
```
[the remark this answers is not in the document]
Comment (isapego) [reply modules/platforms/cpp/common/factory.h:49]: Fixed

File: modules/platforms/cpp/client/include/ignite/table/table.h (hunk 1 above)
Comment (isapego) [reply modules/platforms/cpp/client/include/ignite/table/table.h:42]: No, they should actually still be deleted :) Fixed.

File: modules/platforms/cpp/client/src/cluster_connection.h (hunk 2 above)
Comment (isapego) [reply modules/platforms/cpp/client/src/cluster_connection.h:101]: Fixed.

File: modules/platforms/cpp/protocol/include/ignite/protocol/reader.h (hunk 4 above)
Comment (isapego) [reply modules/platforms/cpp/protocol/include/ignite/protocol/reader.h:125]: Agree. Fixed

File: modules/platforms/cpp/client/include/ignite/table/tables.h [hunk 7]
```diff
@@ -57,12 +57,21 @@ class Tables {
      *   "public.tbl0" - the table "PUBLIC.TBL0" will be looked up,
      *   "PUBLIC.\"Tbl0\"" - "PUBLIC.Tbl0",
      *   "\"MySchema\".\"Tbl0\"" - "MySchema.Tbl0", etc.
-     * @param callback Callback to be called once operation is complete.
-     * @return Table with corresponding name or @c std::nullopt if the table does not exist.
-     * @throw ignite_error In case of error.
+     * @param callback Callback to be called once operation is complete. On success callback called with an instance of
```
[the remark this answers is not in the document]
Comment (isapego) [reply modules/platforms/cpp/client/include/ignite/table/tables.h:60]: Thanks. Fixed.

File: modules/platforms/cpp/client/include/ignite/table/tables.h [hunk 8]
```diff
@@ -57,12 +57,21 @@ class Tables {
...
+     *    the table with corresponding name or @c std::nullopt if the table does not exist.
+     * @throw ignite_error In case of error while trying to send a request.
      */
     IGNITE_API void getTableAsync(const std::string &name, ignite_callback<std::optional<Table>> callback);
 
+    /**
+     * Gets all tables.
+     *
+     * @param callback Callback to be called once operation is complete. On success callback called with a vector of
```
[the remark this answers is not in the document]
Comment (isapego) [reply modules/platforms/cpp/client/include/ignite/table/tables.h:69]: Fixed.

File: modules/platforms/cpp/client/src/table/tables_impl.h (hunk 5 above)
Comment (isapego) [reply modules/platforms/cpp/client/src/table/tables_impl.h:57]: Fixed.

**Pull request merged**